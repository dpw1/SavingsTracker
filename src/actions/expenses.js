import uuid from "uuid";
import database from "../firebase/firebase";

//ADD_EXPENSE
export const addExpense = expense => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const {
      description = "",
      type = "basic-needs",
      note = "",
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = { description, type, note, amount, createdAt };

    database
      .ref(`users/${uid}/expenses`)
      .push(expense)
      .then(ref => {
        dispatch(
          addExpense({
            id: ref.key,
            ...expense
          })
        );
      });
  };
};

export const removeExpense = ({ id } = {}) => ({
  type: "REMOVE_EXPENSE",
  id
});

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .remove()
      .then(() => {
        dispatch(removeExpense({ id }));
      });
  };
};

export const editExpense = (id, updates) => ({
  type: "EDIT_EXPENSE",
  id,
  updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses/${id}`)
      .update({
        ...updates
      })
      .then(() => {
        dispatch(editExpense(id, updates));
      });
  };
};

export const setExpenses = expenses => ({
  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/expenses`)
      .once("value")
      .then(snapshot => {
        let expenses = [];

        snapshot.forEach(child => {
          expenses.push({
            id: child.key,
            ...child.val()
          });
        });

        dispatch(setExpenses(expenses));
      });
  };
};
