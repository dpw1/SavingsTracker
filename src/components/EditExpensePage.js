import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {
  editExpense,
  startRemoveExpense,
  startSetExpenses
} from "../actions/expenses";

const EditExpensePage = props => {
  return (
    <div className="edit-expense">
      <ExpenseForm
        expense={props.expenses}
        onSubmit={expense => {
          props.dispatch(editExpense(props.expenses.id, expense));
          props.history.push("/");
        }}
      />
      <button
        onClick={e => {
          const id = props.expenses.id;
          props.dispatch(startRemoveExpense({ id }));
          props.history.push("/");
        }}
      >
        Remove
      </button>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  return {
    expenses: state.expenses.find(
      expense => expense.id === props.match.params.id
    )
  };
};

export default connect(mapStateToProps)(EditExpensePage);
