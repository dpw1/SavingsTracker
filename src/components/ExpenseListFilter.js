import React from "react";
import { connect } from "react-redux";
import { DateRangePicker } from "react-dates";
import {
  setFilterText,
  sortByAmount,
  sortByDate,
  sortByAlphabetical,
  setStartDate,
  setEndDate
} from "../actions/filters";

class ExpenseListFilter extends React.Component {
  constructor(props) {
    super(props);

    // this.
  }
  state = {
    calendarFocused: null
  };
  onDatesChange = ({ startDate, endDate }) => {
    this.props.dispatch(setStartDate(startDate));
    this.props.dispatch(setEndDate(endDate));
  };
  onFocusChange = calendarFocused => {
    this.setState(() => ({ calendarFocused }));
  };
  render() {
    return (
      <div className="expense-list__filter">
        <input
          type="text"
          value={this.props.filters.text}
          onChange={e => {
            this.props.dispatch(setFilterText(e.target.value));
          }}
        />
        <select
          name=""
          id=""
          onChange={e => {
            switch (e.target.value) {
              case "date":
                return this.props.dispatch(sortByDate());
              case "amount":
                return this.props.dispatch(sortByAmount());
              case "alphabetical":
                return this.props.dispatch(sortByAlphabetical());
              default:
                return "ok";
            }
          }}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
          <option value="alphabetical">Alphabetical</option>
        </select>
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

export default connect(mapStateToProps)(ExpenseListFilter);
