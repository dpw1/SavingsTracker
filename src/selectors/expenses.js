import moment from "moment";

export const getVisibleExpenses = (
  expenses,
  { text, sortBy, sortByLowToHigh, startDate, endDate }
) => {
  return expenses
    .filter(expense => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = RegExp(text.toLowerCase()).test(
        expense.description.toLowerCase()
      );

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      const lowToHigh = sortByLowToHigh === "low-to-high" ? 1 : -1;
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "amount") {
        return a.amount > b.amount ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "type") {
        return a.type > b.type ? lowToHigh * 1 : lowToHigh * -1;
      } else if (sortBy === "alphabetical") {
        return a.description > b.description ? lowToHigh * 1 : lowToHigh * -1;
      }
    });
};
