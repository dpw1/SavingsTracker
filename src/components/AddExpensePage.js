import React from "react";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";
import { connect } from "react-redux";

const AddExpensePage = props => (
  <div className="dashboard">
    <h1>New Expense page!</h1>
    <ExpenseForm
      onSubmit={expense => {
        props.dispatch(startAddExpense(expense));
        props.history.push("/");
      }}
    />
  </div>
);
export default connect()(AddExpensePage);
