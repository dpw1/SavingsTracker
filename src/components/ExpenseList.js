import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import { getVisibleExpenses } from "../selectors/expenses";

export const ExpenseList = props => (
  <div className="expense-list">
    {props.expenses.length === 0 ? (
      <p className="expense-list__empty">No Expenses</p>
    ) : (
      props.expenses.map(expense => {
        return <ExpenseListItem key={expense.id} {...expense} />;
      })
    )}
  </div>
);

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseList);
