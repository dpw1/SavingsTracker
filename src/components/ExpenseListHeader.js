import React from "react";
import ReactDOM from "react-dom";
import { connect } from "react-redux";
import { getVisibleExpenses } from "../selectors/expenses";
import { formatPrice } from "../utils/utils.js";

const ExpenseListHeader = props => {
  return (
    <header className="expense-list__header">
      <h2>
        You're viewing {props.expenses.length}{" "}
        {props.expenses.length === 1 ? "expense" : "expenses"} totalling&nbsp;
        {props.expenses.length >= 1
          ? formatPrice(
              props.expenses.map(each => each.amount).reduce((a, b) => a + b)
            )
          : formatPrice(0)}
      </h2>
    </header>
  );
};

const mapStateToProps = state => {
  return {
    expenses: getVisibleExpenses(state.expenses, state.filters)
  };
};

export default connect(mapStateToProps)(ExpenseListHeader);
