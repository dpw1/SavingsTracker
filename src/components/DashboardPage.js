import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilter from "./ExpenseListFilter";
import ExpenseListHeader from "./ExpenseListHeader";

const DashboardPage = () => (
  <div className="dashboard">
    <ExpenseListHeader />
    <ExpenseListFilter />
    <ExpenseList />
  </div>
);

export default DashboardPage;
