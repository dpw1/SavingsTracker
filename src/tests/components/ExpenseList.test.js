import React from "react";
import { shallow } from "enzyme";
import { ExpenseList } from "../../components/ExpenseList";
import expenses from "../fixtures/expenses";

test("render expense list with expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test("render expense list without expenses", () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
  //   expect(wrapper.find(".expense-list__empty").length).toBe(1);
});
