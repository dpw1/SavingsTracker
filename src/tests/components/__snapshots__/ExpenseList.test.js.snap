// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render expense list with expenses 1`] = `
<div
  className="expense-list"
>
  <Connect(ExpenseListItem)
    amount={100}
    createdAt={"1970-01-01T00:00:00.000Z"}
    description="Gum"
    id="1"
    key="1"
    note=""
    type="charity"
  />
  <Connect(ExpenseListItem)
    amount={200}
    createdAt={-345600000}
    description="Grocery Store"
    id="2"
    key="2"
    note=""
    type="basic-needs"
  />
  <Connect(ExpenseListItem)
    amount={300}
    createdAt={345600000}
    description="Headphone"
    id="3"
    key="3"
    note="20% discount"
    type="miscellaneous"
  />
</div>
`;

exports[`render expense list without expenses 1`] = `
<div
  className="expense-list"
>
  <p
    class="expense-list__empty"
  >
    No Expenses
  </p>
</div>
`;
